### PR TITLE
Add Testerina Service and Client testing guide

### PR DESCRIPTION
**Pull request by Ibaqu**
## Purpose
Guide on how to test Services and Clients in Ballerina

## Check List

- [ ] **Page Addition**
  - [ ] Add `permalink` to pages
  - [ ] If contains empty folder(s), Add front-matter `redirect_to:`

- [ ] **Page Rename**
  - [ ] Add front-matter `redirect_from`
  - [ ] Add front-matter `redirect_to:` (If applicable)

### CODE REVIEW

File: learn/user-guide/testing-ballerina-code/writing-tests.md [hunk 1]
```diff
@@ -665,6 +665,60 @@ variables can be provided in a file named `Config.toml` located in the tests dir
 For information on using configurable variables, see
  [Configurable Variables](/learn/by-example/configurable.html).
 
+Dependencies meant to be resolved only during testing can be specified in the `Ballerina.toml` file
```
Comment (Dilhasha) [inline learn/user-guide/testing-ballerina-code/writing-tests.md:668]: Shall we list it under a topic `Test Dependencies`?

File: learn/user-guide/testing-ballerina-code/writing-tests.md [hunk 2]
```diff
@@ -665,6 +665,60 @@ variables can be provided in a file named `Config.toml` located in the tests dir
...
+
+Testing Ballerina services involves sending specific requests to the service using a client, 
+and verifying the responses using the assertion functions. The aim is to make sure that the service 
+and client behave as expected when sending and recieving both expected requests and malformed ones.
+
+### Configuring Services and Clients
+
+Services or Clients can be defined in such a way as to accommodate for testing under certain 
+conditions by making use of Configurable variables. Maintaining a test configurable configuration 
+allows us to re-define the host or port as required. For example, when defining a Service, the value
```
Comment (Dilhasha) [inline learn/user-guide/testing-ballerina-code/writing-tests.md:687]:
```suggestion
Service or Client configurations can be defined for testing using Configurable variables. Maintaining a test configurable configuration 
allows us to re-define the host or port as required. For example, when defining a Service, the value
```

File: learn/user-guide/testing-ballerina-code/writing-tests.md [hunk 3]
```diff
@@ -665,6 +665,60 @@ variables can be provided in a file named `Config.toml` located in the tests dir
...
+for the host or port can be specified in the `Config.toml` file, which will be used specifically 
+when running the tests.
```
Comment (Dilhasha) [inline learn/user-guide/testing-ballerina-code/writing-tests.md:689]: Let's add a sample Config.toml and mention the path where it needs to be added.

File: learn/user-guide/testing-ballerina-code/writing-tests.md [hunk 4]
```diff
@@ -665,6 +665,60 @@ variables can be provided in a file named `Config.toml` located in the tests dir
...
+
+### Testing Services
+
+Any services defined in the package will start up on their defined ports and will remain running
+for the duration of the testing phase after which they will be automatically shut down. This allows
+you to send requests directly to the service inorder to test its functionality.
+
+To test a service, you can create a client in the test source that sends requests directly to 
+the service and use the test assertion functions to assert the responses. This can be used for both
+services defined in the package as well
```
Comment (Dilhasha) [inline learn/user-guide/testing-ballerina-code/writing-tests.md:710]:
```suggestion
services defined in the package as well.
```

File: learn/user-guide/testing-ballerina-code/writing-tests.md [hunk 5]
```diff
@@ -665,6 +665,60 @@ variables can be provided in a file named `Config.toml` located in the tests dir
...
+    
+} 
+```
+
+### Testing Services
+
+Any services defined in the package will start up on their defined ports and will remain running
+for the duration of the testing phase after which they will be automatically shut down. This allows
+you to send requests directly to the service inorder to test its functionality.
+
```
Comment (Dilhasha) [inline learn/user-guide/testing-ballerina-code/writing-tests.md:707]:
```suggestion
Please note that you need to have tests defined in the respective module where the service is defined in order for the service to start automatically.
```

File: learn/user-guide/testing-ballerina-code/writing-tests.md (hunk 3 above)
Comment (Dilhasha) [reply learn/user-guide/testing-ballerina-code/writing-tests.md:689]: We can provide an example service in the section `Testing Services` and make use of the same example to demonstrate test specific Config.toml. We can reorder the topics to do that. WDYT?

File: learn/user-guide/testing-ballerina-code/writing-tests.md (hunk 1 above)
Comment (praneesha) [inline learn/user-guide/testing-ballerina-code/writing-tests.md:668]:
```suggestion
Dependencies meant to be resolved only during testing and can be specified in the `Ballerina.toml` file
```

File: learn/user-guide/testing-ballerina-code/writing-tests.md [hunk 6]
```diff
@@ -665,6 +665,60 @@ variables can be provided in a file named `Config.toml` located in the tests dir
...
+```toml
+[[platform.java11.dependency]]
+path = "/user/foo/libs/abc.jar"
+scope = "testOnly"
+```
+
+## Testing Ballerina Services and Clients
+
+Testing Ballerina services involves sending specific requests to the service using a client,
```
Comment (praneesha) [inline learn/user-guide/testing-ballerina-code/writing-tests.md:679]:
```suggestion
Testing Ballerina services involves sending specific requests to the service using a client
```

File: learn/user-guide/testing-ballerina-code/writing-tests.md [hunk 7]
```diff
@@ -665,6 +665,60 @@ variables can be provided in a file named `Config.toml` located in the tests dir
...
+## Testing Ballerina Services and Clients
+
+Testing Ballerina services involves sending specific requests to the service using a client, 
+and verifying the responses using the assertion functions. The aim is to make sure that the service 
+and client behave as expected when sending and recieving both expected requests and malformed ones.
+
+### Configuring Services and Clients
+
+Services or Clients can be defined in such a way as to accommodate for testing under certain 
+conditions by making use of Configurable variables. Maintaining a test configurable configuration
```
Comment (praneesha) [inline learn/user-guide/testing-ballerina-code/writing-tests.md:686]:
```suggestion
conditions by making use of configurable variables. Maintaining a test configurable configuration 
```

File: learn/user-guide/testing-ballerina-code/writing-tests.md (hunk 2 above)
Comment (praneesha) [inline learn/user-guide/testing-ballerina-code/writing-tests.md:687]:
```suggestion
allows you to re-define the host or port as required. For example, when defining a service, the value
```

File: learn/user-guide/testing-ballerina-code/writing-tests.md [hunk 8]
```diff
@@ -665,6 +665,60 @@ variables can be provided in a file named `Config.toml` located in the tests dir
...
+service /testService on new http:Listener(port) {
+    
+} 
+```
+
+### Testing Services
+
+Any services defined in the package will start up on their defined ports and will remain running
+for the duration of the testing phase after which they will be automatically shut down. This allows
+you to send requests directly to the service inorder to test its functionality.
```
Comment (praneesha) [inline learn/user-guide/testing-ballerina-code/writing-tests.md:706]:
```suggestion
you to send requests directly to the service in order to test its functionality.
```

File: learn/user-guide/testing-ballerina-code/writing-tests.md (hunk 5 above)
Comment (praneesha) [reply learn/user-guide/testing-ballerina-code/writing-tests.md:707]: Don't use the word "please" in any technical documentation. Can add a note section.

File: learn/user-guide/testing-ballerina-code/writing-tests.md [hunk 9]
```diff
@@ -665,6 +665,60 @@ variables can be provided in a file named `Config.toml` located in the tests dir
...
+
+To test a service, you can create a client in the test source that sends requests directly to
```
Comment (praneesha) [inline learn/user-guide/testing-ballerina-code/writing-tests.md:708]:
```suggestion
To test a service, you can create a client in the test source, which sends requests directly to 
```

File: learn/user-guide/testing-ballerina-code/writing-tests.md [hunk 10]
```diff
@@ -665,6 +665,60 @@ variables can be provided in a file named `Config.toml` located in the tests dir
...
+for the duration of the testing phase after which they will be automatically shut down. This allows
+you to send requests directly to the service inorder to test its functionality.
+
+To test a service, you can create a client in the test source that sends requests directly to 
+the service and use the test assertion functions to assert the responses. This can be used for both
+services defined in the package as well 
+
+### Testing Clients
+
+In cases where a fully fledged client is already defined for a particular service, we can make use
```
Comment (praneesha) [inline learn/user-guide/testing-ballerina-code/writing-tests.md:714]:
```suggestion
In cases where a fully fledged client is already defined for a particular service, you can make use 
```

File: learn/user-guide/testing-ballerina-code/writing-tests.md [hunk 11]
```diff
@@ -665,6 +665,60 @@ variables can be provided in a file named `Config.toml` located in the tests dir
...
+you to send requests directly to the service inorder to test its functionality.
+
+To test a service, you can create a client in the test source that sends requests directly to 
+the service and use the test assertion functions to assert the responses. This can be used for both
+services defined in the package as well 
+
+### Testing Clients
+
+In cases where a fully fledged client is already defined for a particular service, we can make use 
+of Object Mocking to mock the calls to the service and return curated responses to the client.
```
Comment (praneesha) [inline learn/user-guide/testing-ballerina-code/writing-tests.md:715]:
```suggestion
of object mocking to mock the calls to the service and return curated responses to the client.
```

File: learn/user-guide/testing-ballerina-code/writing-tests.md (hunk 1 above)
Comment (Ibaqu) [reply learn/user-guide/testing-ballerina-code/writing-tests.md]: Created a new Topic `Test Dependencies`

File: learn/user-guide/testing-ballerina-code/writing-tests.md (hunk 3 above)
Comment (Ibaqu) [reply learn/user-guide/testing-ballerina-code/writing-tests.md]: Added an example and moved the configuration portion towards the end of the guide

**Review by Dilhasha — APPROVED** [review]

**Pull request merged**